[API] ImGuizmo

AAstroPhysiCS opened the pull request:
This PR finalizes #41 and adds ImGuizmo.

I've already added an example for you to try. There are random weird crashes with the snapping feature and there's still a lot to add, so do not merge it yet.
I am gonna let you know, after I've fixed the crashes and added more features, so that you can go ahead and merge it.
Please have a quick look at the code and let me know if I have any errors/type styles to fix.

Example:

https://user-images.githubusercontent.com/25062570/117713749-590bdf00-b1d6-11eb-8e4a-6c6425cc992a.mp4

SpaiR commented:
Hi. Good work! Appreciate the example introduction.
It's problematic to do a proper review at the moment, since your diff is to "noisy". Specifically:
 - removed `bin/` dir
 - `imgui-binding/tmp/` dir
 - `libsNative/` dir
 
 That all create a 65k line of changes and I don't know where to look for real changes. Please check the original PR. I made a small review of it, I see the same probs here as well.
 
 GIF with example looks good.

AAstroPhysiCS commented:
I've updated the example, deleted the unnecessary build files and fixed crashes.

One thing that I want to note here, is that I am not gonna implement any other features such as ImSequencer etc.
For me, ImGuizmo is enough for now.

I've added comments for you to look and give feedback (in the example file: ExampleImGuizmo.java)

Thanks

SpaiR commented on example/src/main/java/ExampleImGuizmo.java:
```diff
@@ -0,0 +1,289 @@
+            ImGui.text(ImGuizmo.isOver(Operation.TRANSLATE) ? "Over translate gizmo" : "");
+            ImGui.text(ImGuizmo.isOver(Operation.ROTATE) ? "Over rotate gizmo" : "");
+            ImGui.text(ImGuizmo.isOver(Operation.SCALE) ? "Over scale gizmo" : "");
+        } else ImGui.text("Not using gizmo");
```
Follow general code style and do not do one line `if ... else ...` statements.

SpaiR commented on example/src/main/java/ExampleImGuizmo.java:
```diff
@@ -0,0 +1,289 @@
+
+    private static void editTransform() {
+
+        if (ImGui.isKeyPressed(90)) currentGizmoOperation = Operation.TRANSLATE;
```
Same thing with one-liners. Also, use `GLFW.GLFW_KEY_*` constants instead of raw numbers to express pressed key.

SpaiR commented on example/src/main/java/ExampleImGuizmo.java:
```diff
@@ -0,0 +1,289 @@
+        ImGui.inputFloat3("Rt", matrixRotation, "%.3f", ImGuiInputTextFlags.ReadOnly);
+        ImGui.inputFloat3("Sc", matrixScale, "%.3f", ImGuiInputTextFlags.ReadOnly);
+
+        if (ImGuizmo.isUsing())
```
Here and below, use `{}` for every `if` statement.

SpaiR commented on imgui-app/src/main/java/imgui/app/Window.java:
```diff
@@ -163,6 +164,7 @@ protected void startFrame() {
         GL32.glClear(GL32.GL_COLOR_BUFFER_BIT | GL32.GL_DEPTH_BUFFER_BIT);
         imGuiGlfw.newFrame();
         ImGui.newFrame();
+        ImGuizmo.beginFrame();
```
`Window` is a general class to implement ImGui API. So other stuff like extension usage should be implemented in user specific code. Method `startFrame` could be overriden with additional logic. This line should be removed from here.

SpaiR commented on imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java:
```diff
@@ -0,0 +1,276 @@
+
+public final class ImGuizmo {
+
+    private static float[] matrices = null; //for drawCubes()
```
Maybe create array by default and make variable final?

AAstroPhysiCS replied:
No, we don't want unnecessary allocations. If we'd make it final, java would reserve 64 floats, which corresponds to 256 bytes even, if the user does not want to render debugging cubes on the screen.

The allocation should only happen, if the user really wants to render debugging cubes on the screen, hence the allocation in the drawCubes() method.

SpaiR commented on imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java:
```diff
@@ -0,0 +1,276 @@
+
+     /*JNI
+       #include "_common.h"
+       #include "ImGuizmo.h"
```
Use `_imguizmo.h` header instead.

SpaiR commented on imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java:
```diff
@@ -0,0 +1,276 @@
+    }
+
+    private static native boolean nIsUsing(); /*
+        return (jboolean) ImGuizmo::IsUsing();
```
You don't need to cast to `jboolean` anywhere.

SpaiR commented on imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java:
```diff
@@ -0,0 +1,276 @@
+        ImGuizmo::DecomposeMatrixToComponents(matrix, translation, rotation, scale);
+    */
+
+    /*
+     * helper functions for manualy editing translation/rotation/scale with an input float
+     * translation, rotation and scale float points to 3 floats each
+     * Angles are in degrees (more suitable for human editing)
+     * example:
+     * float matrixTranslation[3], matrixRotation[3], matrixScale[3];
+     * ImGuizmo::DecomposeMatrixToComponents(gizmoMatrix.m16, matrixTranslation, matrixRotation, matrixScale);
+     * ImGui::InputFloat3("Tr", matrixTranslation, 3);
+     * ImGui::InputFloat3("Rt", matrixRotation, 3);
+     * ImGui::InputFloat3("Sc", matrixScale, 3);
+     * ImGuizmo::RecomposeMatrixFromComponents(matrixTranslation, matrixRotation, matrixScale, gizmoMatrix.m16);
+     * These functions have some numerical stability issues for now. Use with caution.
+     */
```
Can you make it a proppert javadoc? I do release it as well.

SpaiR commented on imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java:
```diff
@@ -0,0 +1,276 @@
+     */
+    public static void drawCubes(final float[] view, final float[] projection, final float[]... cubeMatrices) {
+        if (cubeMatrices.length > 4) {
+            System.err.println("Drawing cubes with ImGuizmo only supports up to 4 cubes because it should only be used for debugging purposes");
```
Throw `IllegalArgumentException` instead to make error more exlicit.

SpaiR reviewed (changes requested):
And, as I said, please remove `bin/` folder.

AAstroPhysiCS commented:
I've did more and included pretty much doc for all the necessary methods. It looks professional now and awaits for your approval. Hit me up if any changes are necessary.

AAstroPhysiCS commented:
The check failed because there's a unused import in the ImGuizmo class but ImGuizmo class **_uses_** the import for the comments. See `ImGuizmo.java 317:71`
So I guess, your check doesn't check if the import is being used somewhere in the comments.

Other than that, everything checks out.

SpaiR reviewed (commented):
By removing the bin folder I meant to remove it from your diff, not to delete completle. Just revert all changes to the dir.

> The check failed because there's a unused import in the ImGuizmo class but ImGuizmo class uses the import for the comment.

Just do the import in the javadoc itself. Like: `{@link imgui.extension.imguizmo.flag.Operation#ROTATE}`

AAstroPhysiCS commented:
I've reverted all changes to the `bin/` folder. Reverted all the changes back. Should be ready now.

SpaiR commented:
For your info, I'll merge the PR on the weekend. Dear ImGui also had released 1.83, so I'll do everything in one go. If you have something to polish - there is time to do that.

AAstroPhysiCS commented:
I see that you force pushed your own commits in to my master branch. Can I get the reason why you are doing this, since force pushing can produce a lot of mayhem. Additionally, my commits are "deleted" since you force pushed it.

SpaiR commented:
I prepared the PR for the merge. Your commits are not "deleted", but squashed. I would do it anyway during the merge itself. 

Force push doesn't produce a mayhem, it's a normal practice to squash a bunch of commits into one, since it's an atomic change. 
Rebases in general are good too. Using of master to commit changes from fork is the bad practice though. I would recommend you to avoid that in future and create a separte branch for you stuff.
